File: leetcode/medium/0662MaxWidth/maxwidth.cpp
#include "bits/stdc++.h";

using namespace std;

//Definition for a binary tree node.
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution
{
public:
    int widthOfBinaryTree(TreeNode *root)
    {
        if (root == nullptr)
            return 0;
        dfs(root, 0, 0);
        return maxWidth;
    }

private:
    int maxWidth = INT_MIN;
    unordered_map<int, unsigned long int> leftPosition;
    void dfs(TreeNode *root, unsigned long int position, int depth)
    {
        if (root == nullptr)
            return;

        if (leftPosition.count(depth) == 0)
        {
            leftPosition[depth] = position;
        }

        int leftMostposition = leftPosition[depth];
        /*Calculate the max width*/
        int widthAtCurrentDepth = position - leftMostposition + 1;
        maxWidth = max(maxWidth, widthAtCurrentDepth);

        dfs(root->left, position * 2, depth + 1);
        dfs(root->right, position * 2 + 1, depth + 1);
    }
};

File: leetcode/medium/0662MaxWidth/maxwidth.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def __init__(self):
        self.left = {}
        self.max_width = 0

    def widthOfBinaryTree(self, root):    
        self.dfs(root, 0, 0)
        return self.max_width
    def dfs(self, root, depth, position):
        if root is None:
            return
        if depth not in self.left:
            self.left[depth] = position
        self.max_width = max(self.max_width, position - self.left[depth] + 1)
        self.dfs(root.left, depth + 1, 2 * position)
        self.dfs(root.right, depth + 1, 2 * position + 1)

